File: SundeeFundee/Sources/SundeeFundeeKit/DomainLayer/Intelligence/PlateauDetector.swift
import Foundation

// MARK: - PlateauDetector

/// Detects stalled lifts from 1RM history and generates actionable alerts.
///
/// A plateau is defined as a lift whose 1RM has not increased over a
/// configurable window of attempts. The detector groups records by exercise,
/// sorts chronologically, and checks whether recent attempts show progress.
///
/// Pure domain logic — no framework dependencies.
public enum PlateauDetector {

    // MARK: - Configuration

    /// Minimum number of records required before a plateau can be detected.
    public static let minimumAttempts = 3

    /// Number of most-recent records to evaluate for progress.
    public static let recentWindow = 3

    /// Maximum allowed days between first and last attempt in the window
    /// for a plateau to be flagged. Prevents stale data from triggering alerts.
    public static let maxWindowDays = 90

    /// Default percentage improvement required to count as progress.
    /// 0.01 = 1% improvement over the window.
    public static let progressThreshold = 0.01

    /// Minimum records needed for rate-of-progress analysis.
    public static let minimumProgressRecords = 5

    // MARK: - Types

    /// Whether the plateau is a strength (1RM) or volume stall.
    public enum PlateauType: String, Sendable, Equatable {
        case strength
        case volume
    }

    /// A detected plateau for a specific lift.
    public struct PlateauAlert: Sendable, Equatable {
        /// The exercise name that has stalled.
        public let exerciseName: String

        /// The current 1RM weight (most recent).
        public let currentWeight: Double

        /// The weight unit.
        public let unit: WeightUnit

        /// The best weight achieved in the evaluation window.
        public let bestWeight: Double

        /// Number of consecutive attempts without meaningful progress.
        public let stallCount: Int

        /// Days since the best weight was achieved.
        public let daysSinceBest: Int

        /// Actionable recommendation.
        public let recommendation: String

        /// Whether this is a strength or volume plateau.
        public let plateauType: PlateauType

        public init(
            exerciseName: String,
            currentWeight: Double,
            unit: WeightUnit,
            bestWeight: Double,
            stallCount: Int,
            daysSinceBest: Int,
            recommendation: String,
            plateauType: PlateauType = .strength
        ) {
            self.exerciseName = exerciseName
            self.currentWeight = currentWeight
            self.unit = unit
            self.bestWeight = bestWeight
            self.stallCount = stallCount
            self.daysSinceBest = daysSinceBest
            self.recommendation = recommendation
            self.plateauType = plateauType
        }
    }

    /// An early warning that progress is slowing before a full plateau.
    public struct RateOfProgressAlert: Sendable, Equatable {
        public let exerciseName: String
        /// Weight gained per week in the recent period.
        public let currentRate: Double
        /// Weight gained per week in the earlier period.
        public let historicalRate: Double
        public let message: String
    }

    // MARK: - Exercise-Specific Thresholds

    /// Returns an exercise-category-aware progress threshold.
    /// Compound lifts move larger absolute weights and progress more slowly,
    /// so they use a tighter threshold. Isolation/press exercises use a wider one.
    public static func progressThreshold(for category: WeightliftingCategory) -> Double {
        switch category {
        case .squat, .hipHinge, .olympicWeightlifting:
            return 0.005  // 0.5% — compound lifts progress slowly
        case .press, .pull:
            return 0.015  // 1.5% — upper body lifts progress faster
        case .carry:
            return 0.01   // 1.0% — default
        }
    }

    /// Resolves the category for an exercise name from the catalog, falling back to nil.
    private static func resolveCategory(
        exerciseName: String,
        catalog: [WeightliftingEntry]
    ) -> WeightliftingCategory? {
        catalog.first { $0.id == exerciseName }?.category
    }

    // MARK: - Strength Plateau Detection

    /// Analyzes 1RM records and returns alerts for stalled lifts.
    ///
    /// - Parameters:
    ///   - records: All 1RM records for the user, across all exercises.
    ///   - catalog: The exercise catalog for category-aware thresholds. When provided,
    ///     compound lifts use a tighter threshold (0.5%) while isolation lifts use a wider one (1.5%).
    /// - Returns: An array of `PlateauAlert` for exercises that appear stalled.
    public static func detect(
        from records: [OneRepMaxRecord],
        catalog: [WeightliftingEntry] = []
    ) -> [PlateauAlert] {
        let grouped = Dictionary(grouping: records) { $0.exerciseName }

        var alerts: [PlateauAlert] = []

        for (exerciseName, exerciseRecords) in grouped {
            guard exerciseRecords.count >= minimumAttempts else { continue }

            let sorted = exerciseRecords.sorted { $0.date < $1.date }

            let recent = Array(sorted.suffix(recentWindow))
            guard recent.count >= minimumAttempts else { continue }

            guard let first = recent.first, let last = recent.last else { continue }
            let daySpan = Calendar.current.dateComponents([.day], from: first.date, to: last.date).day ?? 0
            guard daySpan <= maxWindowDays else { continue }

            let bestWeight = recent.map { $0.weight }.max() ?? 0
            let currentWeight = last.weight

            let oldestInWindow = first.weight
            let improvement = oldestInWindow > 0
                ? (bestWeight - oldestInWindow) / oldestInWindow
                : 0

            // Use category-specific threshold when catalog is available
            let threshold: Double
            if let category = resolveCategory(exerciseName: exerciseName, catalog: catalog) {
                threshold = progressThreshold(for: category)
            } else {
                threshold = progressThreshold
            }

            if improvement < threshold {
                let peak = sorted.map { $0.weight }.max() ?? 0
                var stallCount = 0
                var seenPeak = false
                for record in sorted.reversed() {
                    if record.weight >= peak && !seenPeak {
                        seenPeak = true
                        stallCount += 1
                    } else if seenPeak && record.weight >= peak {
                        stallCount += 1
                    } else if seenPeak {
                        break
                    } else {
                        stallCount += 1
                    }
                }
                stallCount = max(stallCount, recent.count)

                let bestDate = sorted.last(where: { $0.weight == peak })?.date ?? last.date
                let daysSinceBest = Calendar.current.dateComponents(
                    [.day], from: bestDate, to: Date()
                ).day ?? 0

                let recommendation = generateRecommendation(
                    exerciseName: exerciseName,
                    stallCount: stallCount,
                    currentWeight: currentWeight,
                    bestWeight: peak,
                    plateauType: .strength
                )

                alerts.append(PlateauAlert(
                    exerciseName: exerciseName,
                    currentWeight: currentWeight,
                    unit: last.unit,
                    bestWeight: peak,
                    stallCount: stallCount,
                    daysSinceBest: daysSinceBest,
                    recommendation: recommendation,
                    plateauType: .strength
                ))
            }
        }

        return alerts.sorted { $0.stallCount > $1.stallCount }
    }

    // MARK: - Volume Plateau Detection

    /// Detects exercises whose training volume has stalled over recent weeks.
    ///
    /// Groups workouts by week, computes per-exercise weekly volume (reps x weight),
    /// and flags exercises with less than 2% volume growth across the window.
    ///
    /// - Parameters:
    ///   - workouts: All completed workouts.
    ///   - windowWeeks: Number of weeks to evaluate (default 4).
    /// - Returns: Alerts for exercises with stalled volume.
    public static func detectVolumePlateaus(
        from workouts: [Workout],
        windowWeeks: Int = 4
    ) -> [PlateauAlert] {
        let calendar = Calendar.current
        let cutoffDate = calendar.date(byAdding: .weekOfYear, value: -windowWeeks, to: Date()) ?? Date()
        let recentWorkouts = workouts.filter { $0.date >= cutoffDate && $0.isComplete }

        guard recentWorkouts.count >= 2 else { return [] }

        // Group workouts by week number
        var weeklyVolume: [Int: [String: Double]] = [:]
        for workout in recentWorkouts {
            let weekNumber = calendar.component(.weekOfYear, from: workout.date)
            let yearWeek = calendar.component(.year, from: workout.date) * 100 + weekNumber
            for exercise in workout.exercises {
                let volume = exercise.targetSets.reduce(0.0) { sum, set in
                    let reps = set.actualReps ?? set.reps
                    let weight = (set.completedWeight ?? 0) > 0
                        ? set.completedWeight!
                        : (set.prescribedWeight > 0 ? set.prescribedWeight : 0)
                    return sum + Double(reps) * weight
                }
                weeklyVolume[yearWeek, default: [:]][exercise.name, default: 0] += volume
            }
        }

        guard weeklyVolume.count >= 2 else { return [] }

        let sortedWeeks = weeklyVolume.keys.sorted()
        // Collect all exercises that appear in at least 2 weeks
        var allExercises = Set<String>()
        for (_, exercises) in weeklyVolume {
            allExercises.formUnion(exercises.keys)
        }

        var alerts: [PlateauAlert] = []
        let volumeGrowthThreshold = 0.02 // 2%

        for exerciseName in allExercises {
            let volumes = sortedWeeks.compactMap { weeklyVolume[$0]?[exerciseName] }
            guard volumes.count >= 2 else { continue }

            let firstVolume = volumes.first!
            let lastVolume = volumes.last!
            guard firstVolume > 0 else { continue }

            let growth = (lastVolume - firstVolume) / firstVolume
            if growth < volumeGrowthThreshold {
                let recommendation = generateRecommendation(
                    exerciseName: exerciseName,
                    stallCount: volumes.count,
                    currentWeight: lastVolume,
                    bestWeight: volumes.max() ?? lastVolume,
                    plateauType: .volume
                )
                alerts.append(PlateauAlert(
                    exerciseName: exerciseName,
                    currentWeight: lastVolume,
                    unit: .lbs,
                    bestWeight: volumes.max() ?? lastVolume,
                    stallCount: volumes.count,
                    daysSinceBest: 0,
                    recommendation: recommendation,
                    plateauType: .volume
                ))
            }
        }

        return alerts.sorted { $0.stallCount > $1.stallCount }
    }

    // MARK: - Rate of Progress Detection

    /// Detects exercises where the rate of strength progress is slowing,
    /// providing an early warning before a full plateau manifests.
    ///
    /// Splits 1RM history into two halves and compares the rate of weight
    /// gain per week. If the recent rate drops below 50% of the earlier rate,
    /// an alert is generated.
    ///
    /// - Parameters:
    ///   - records: All 1RM records for the user.
    ///   - minimumRecords: Minimum records per exercise (default 5).
    /// - Returns: Early warning alerts for slowing progress.
    public static func detectSlowingProgress(
        from records: [OneRepMaxRecord],
        minimumRecords: Int = 5
    ) -> [RateOfProgressAlert] {
        let grouped = Dictionary(grouping: records) { $0.exerciseName }
        var alerts: [RateOfProgressAlert] = []

        for (exerciseName, exerciseRecords) in grouped {
            guard exerciseRecords.count >= minimumRecords else { continue }

            let sorted = exerciseRecords.sorted { $0.date < $1.date }

            // Split into first half and second half
            let midpoint = sorted.count / 2
            let firstHalf = Array(sorted.prefix(midpoint))
            let secondHalf = Array(sorted.suffix(from: midpoint))

            guard firstHalf.count >= 2, secondHalf.count >= 2 else { continue }

            let historicalRate = ratePerWeek(records: firstHalf)
            let currentRate = ratePerWeek(records: secondHalf)

            // Only flag if historical rate was positive (they were progressing)
            // and current rate has dropped below 50% of historical
            if historicalRate > 0 && currentRate < historicalRate * 0.5 {
                let message: String
                if currentRate <= 0 {
                    message = "Your progress on \(exerciseName) has stalled — consider a variation or deload before a full plateau."
                } else {
                    message = "Your progress on \(exerciseName) is slowing — consider a variation or deload before a full plateau."
                }
                alerts.append(RateOfProgressAlert(
                    exerciseName: exerciseName,
                    currentRate: currentRate,
                    historicalRate: historicalRate,
                    message: message
                ))
            }
        }

        return alerts
    }

    /// Computes weight gained per week across a sorted slice of records.
    private static func ratePerWeek(records: [OneRepMaxRecord]) -> Double {
        guard let first = records.first, let last = records.last else { return 0 }
        let weeks = Calendar.current.dateComponents([.day], from: first.date, to: last.date).day.map { Double($0) / 7.0 } ?? 0
        guard weeks > 0 else { return 0 }
        return (last.weight - first.weight) / weeks
    }

    // MARK: - Recommendations

    private static func generateRecommendation(
        exerciseName: String,
        stallCount: Int,
        currentWeight: Double,
        bestWeight: Double,
        plateauType: PlateauType
    ) -> String {
        switch plateauType {
        case .strength:
            if stallCount >= 6 {
                return "Consider changing your rep scheme or adding a variation of \(exerciseName) for 3-4 weeks, then re-test."
            } else if stallCount >= 4 {
                return "Try a deload week at 60% of your \(exerciseName) max, then build back with smaller jumps."
            } else {
                return "Add volume: try 3-5 extra sets of \(exerciseName) per week at 70-80% to break through."
            }
        case .volume:
            return "Your \(exerciseName) volume has plateaued. Try adding 1-2 sets per session or increase weight by 5%."
        }
    }
}

File: android/core/src/main/java/com/sundeefundee/core/domain/intelligence/PlateauDetector.kt
package com.sundeefundee.core.domain.intelligence

import com.sundeefundee.core.domain.exercise.WeightliftingCategory
import com.sundeefundee.core.domain.exercise.WeightliftingCatalog
import com.sundeefundee.core.model.OneRepMaxRecord
import com.sundeefundee.core.model.WeightUnit
import com.sundeefundee.core.model.Workout
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.temporal.ChronoUnit
import kotlin.math.max
import kotlin.math.round

/**
 * Detects stalled lifts from 1RM history and generates actionable alerts.
 *
 * A plateau is defined as a lift whose 1RM has not increased over a
 * configurable window of attempts. The detector groups records by exercise,
 * sorts chronologically, and checks whether recent attempts show progress.
 *
 * Pure domain logic -- no framework dependencies.
 */
object PlateauDetector {

    // MARK: - Configuration

    /** Minimum number of records required before a plateau can be detected. */
    const val MINIMUM_ATTEMPTS = 3

    /** Number of most-recent records to evaluate for progress. */
    const val RECENT_WINDOW = 3

    /** Maximum allowed days between first and last attempt in the window
     * for a plateau to be flagged. Prevents stale data from triggering alerts. */
    const val MAX_WINDOW_DAYS = 90

    /** Default percentage improvement required to count as progress.
     * 0.01 = 1% improvement over the window. */
    const val PROGRESS_THRESHOLD = 0.01

    /** Minimum records needed for rate-of-progress analysis. */
    const val MINIMUM_PROGRESS_RECORDS = 5

    // MARK: - Types

    /** Whether the plateau is a strength (1RM) or volume stall. */
    enum class PlateauType {
        STRENGTH,
        VOLUME
    }

    /** A detected plateau for a specific lift. */
    data class PlateauAlert(
        /** The exercise name that has stalled. */
        val exerciseName: String,
        /** The current 1RM weight (most recent). */
        val currentWeight: Double,
        /** The weight unit. */
        val unit: WeightUnit,
        /** The best weight achieved in the evaluation window. */
        val bestWeight: Double,
        /** Number of consecutive attempts without meaningful progress. */
        val stallCount: Int,
        /** Days since the best weight was achieved. */
        val daysSinceBest: Int,
        /** Actionable recommendation. */
        val recommendation: String,
        /** Whether this is a strength or volume plateau. */
        val plateauType: PlateauType = PlateauType.STRENGTH
    )

    /** An early warning that progress is slowing before a full plateau. */
    data class RateOfProgressAlert(
        val exerciseName: String,
        /** Weight gained per week in the recent period. */
        val currentRate: Double,
        /** Weight gained per week in the earlier period. */
        val historicalRate: Double,
        val message: String
    )

    // MARK: - Exercise-Specific Thresholds

    /** Returns an exercise-category-aware progress threshold.
     * Compound lifts move larger absolute weights and progress more slowly,
     * so they use a tighter threshold. Isolation/press exercises use a wider one. */
    fun progressThresholdForCategory(category: WeightliftingCategory): Double {
        return when (category) {
            WeightliftingCategory.SQUAT,
            WeightliftingCategory.HIP_HINGE,
            WeightliftingCategory.OLYMPIC_WEIGHTLIFTING -> 0.005 // 0.5%

            WeightliftingCategory.PRESS,
            WeightliftingCategory.PULL -> 0.015 // 1.5%

            WeightliftingCategory.CARRY -> 0.01 // 1.0%
        }
    }

    /** Resolves the category for an exercise name from the catalog, falling back to null. */
    private fun resolveCategory(exerciseName: String): WeightliftingCategory? {
        return WeightliftingCatalog.all.firstOrNull { it.id == exerciseName }?.category
    }

    // MARK: - Strength Plateau Detection

    /**
     * Analyzes 1RM records and returns alerts for stalled lifts.
     *
     * @param records All 1RM records for the user, across all exercises.
     * @return An array of [PlateauAlert] for exercises that appear stalled.
     */
    fun detect(records: List<OneRepMaxRecord>): List<PlateauAlert> {
        val grouped = records.groupBy { it.exerciseName }

        val alerts = mutableListOf<PlateauAlert>()

        for ((exerciseName, exerciseRecords) in grouped) {
            if (exerciseRecords.size < MINIMUM_ATTEMPTS) continue

            val sorted = exerciseRecords.sortedBy { parseDate(it.date) }
            val recent = sorted.takeLast(RECENT_WINDOW)
            if (recent.size < MINIMUM_ATTEMPTS) continue

            val first = recent.first()
            val last = recent.last()

            val daySpan = ChronoUnit.DAYS.between(
                parseDate(first.date), parseDate(last.date)
            )
            if (daySpan > MAX_WINDOW_DAYS) continue

            val bestWeight = recent.maxOf { it.weight }
            val currentWeight = last.weight

            val oldestInWindow = first.weight
            val improvement = if (oldestInWindow > 0) {
                (bestWeight - oldestInWindow) / oldestInWindow
            } else {
                0.0
            }

            // Use category-specific threshold when catalog is available
            val threshold: Double = resolveCategory(exerciseName)?.let {
                progressThresholdForCategory(it)
            } ?: PROGRESS_THRESHOLD

            if (improvement < threshold) {
                val peak = sorted.maxOf { it.weight }
                var stallCount = 0
                var seenPeak = false

                for (record in sorted.reversed()) {
                    if (record.weight >= peak && !seenPeak) {
                        seenPeak = true
                        stallCount++
                    } else if (seenPeak && record.weight >= peak) {
                        stallCount++
                    } else if (seenPeak) {
                        break
                    } else {
                        stallCount++
                    }
                }
                stallCount = max(stallCount, recent.size)

                val bestRecord = sorted.lastOrNull { it.weight == peak }
                val bestDate = bestRecord?.let { parseDate(it.date) } ?: parseDate(last.date)
                val daysSinceBest = ChronoUnit.DAYS.between(bestDate, LocalDate.now()).toInt()

                val recommendation = generateRecommendation(
                    exerciseName = exerciseName,
                    stallCount = stallCount,
                    currentWeight = currentWeight,
                    bestWeight = peak,
                    plateauType = PlateauType.STRENGTH
                )

                alerts.add(
                    PlateauAlert(
                        exerciseName = exerciseName,
                        currentWeight = currentWeight,
                        unit = last.unit,
                        bestWeight = peak,
                        stallCount = stallCount,
                        daysSinceBest = daysSinceBest,
                        recommendation = recommendation,
                        plateauType = PlateauType.STRENGTH
                    )
                )
            }
        }

        return alerts.sortedByDescending { it.stallCount }
    }

    // MARK: - Volume Plateau Detection

    /**
     * Detects exercises whose training volume has stalled over recent weeks.
     *
     * Groups workouts by week, computes per-exercise weekly volume (reps x weight),
     * and flags exercises with less than 2% volume growth across the window.
     *
     * @param workouts All completed workouts.
     * @param windowWeeks Number of weeks to evaluate (default 4).
     * @return Alerts for exercises with stalled volume.
     */
    fun detectVolumePlateaus(
        workouts: List<Workout>,
        windowWeeks: Int = 4
    ): List<PlateauAlert> {
        val cutoffDate = LocalDate.now().minusWeeks(windowWeeks.toLong())
        val recentWorkouts = workouts.filter {
            parseDate(it.date) >= cutoffDate && it.isComplete
        }

        if (recentWorkouts.size < 2) return emptyList()

        // Group workouts by year*100 + week number
        val weeklyVolume = mutableMapOf<Int, MutableMap<String, Double>>()

        for (workout in recentWorkouts) {
            val workoutDate = parseDate(workout.date)
            val yearWeek = workoutDate.year * 100 + workoutDate.get(
                java.time.temporal.WeekFields.ISO.weekOfWeekBasedYear()
            )

            for (exercise in workout.exercises) {
                val volume = exercise.targetSets.sumOf { set ->
                    val reps = (set.actualReps ?: set.reps).toDouble()
                    val weight = if ((set.completedWeight ?: 0.0) > 0) {
                        set.completedWeight!!
                    } else if (set.prescribedWeight > 0) {
                        set.prescribedWeight
                    } else {
                        0.0
                    }
                    reps * weight
                }

                weeklyVolume
                    .getOrPut(yearWeek) { mutableMapOf() }
                    .merge(exercise.name, volume) { old, new -> old + new }
            }
        }

        if (weeklyVolume.size < 2) return emptyList()

        val sortedWeeks = weeklyVolume.keys.sorted()

        // Collect all exercises that appear in at least 2 weeks
        val allExercises = mutableSetOf<String>()
        for ((_, exercises) in weeklyVolume) {
            allExercises.addAll(exercises.keys)
        }

        val alerts = mutableListOf<PlateauAlert>()
        val volumeGrowthThreshold = 0.02 // 2%

        for (exerciseName in allExercises) {
            val volumes = sortedWeeks.mapNotNull { weeklyVolume[it]?.get(exerciseName) }
            if (volumes.size < 2) continue

            val firstVolume = volumes.first()
            val lastVolume = volumes.last()
            if (firstVolume <= 0) continue

            val growth = (lastVolume - firstVolume) / firstVolume
            if (growth < volumeGrowthThreshold) {
                val bestVolume = volumes.max()
                val recommendation = generateRecommendation(
                    exerciseName = exerciseName,
                    stallCount = volumes.size,
                    currentWeight = lastVolume,
                    bestWeight = bestVolume,
                    plateauType = PlateauType.VOLUME
                )
                alerts.add(
                    PlateauAlert(
                        exerciseName = exerciseName,
                        currentWeight = lastVolume,
                        unit = WeightUnit.LBS,
                        bestWeight = bestVolume,
                        stallCount = volumes.size,
                        daysSinceBest = 0,
                        recommendation = recommendation,
                        plateauType = PlateauType.VOLUME
                    )
                )
            }
        }

        return alerts.sortedByDescending { it.stallCount }
    }

    // MARK: - Rate of Progress Detection

    /**
     * Detects exercises where the rate of strength progress is slowing,
     * providing an early warning before a full plateau manifests.
     *
     * Splits 1RM history into two halves and compares the rate of weight
     * gain per week. If the recent rate drops below 50% of the earlier rate,
     * an alert is generated.
     *
     * @param records All 1RM records for the user.
     * @param minimumRecords Minimum records per exercise (default 5).
     * @return Early warning alerts for slowing progress.
     */
    fun detectSlowingProgress(
        records: List<OneRepMaxRecord>,
        minimumRecords: Int = 5
    ): List<RateOfProgressAlert> {
        val grouped = records.groupBy { it.exerciseName }
        val alerts = mutableListOf<RateOfProgressAlert>()

        for ((exerciseName, exerciseRecords) in grouped) {
            if (exerciseRecords.size < minimumRecords) continue

            val sorted = exerciseRecords.sortedBy { parseDate(it.date) }

            // Split into first half and second half
            val midpoint = sorted.size / 2
            val firstHalf = sorted.take(midpoint)
            val secondHalf = sorted.drop(midpoint)

            if (firstHalf.size < 2 || secondHalf.size < 2) continue

            val historicalRate = ratePerWeek(firstHalf)
            val currentRate = ratePerWeek(secondHalf)

            // Only flag if historical rate was positive (they were progressing)
            // and current rate has dropped below 50% of historical
            if (historicalRate > 0 && currentRate < historicalRate * 0.5) {
                val message = if (currentRate <= 0) {
                    "Your progress on $exerciseName has stalled -- consider a variation or deload before a full plateau."
                } else {
                    "Your progress on $exerciseName is slowing -- consider a variation or deload before a full plateau."
                }
                alerts.add(
                    RateOfProgressAlert(
                        exerciseName = exerciseName,
                        currentRate = currentRate,
                        historicalRate = historicalRate,
                        message = message
                    )
                )
            }
        }

        return alerts
    }

    // MARK: - Private Helpers

    /** Computes weight gained per week across a sorted slice of records. */
    private fun ratePerWeek(records: List<OneRepMaxRecord>): Double {
        val first = records.firstOrNull() ?: return 0.0
        val last = records.lastOrNull() ?: return 0.0

        val days = ChronoUnit.DAYS.between(parseDate(first.date), parseDate(last.date)).toDouble()
        val weeks = days / 7.0
        if (weeks <= 0) return 0.0

        return (last.weight - first.weight) / weeks
    }

    private fun generateRecommendation(
        exerciseName: String,
        stallCount: Int,
        currentWeight: Double,
        bestWeight: Double,
        plateauType: PlateauType
    ): String {
        return when (plateauType) {
            PlateauType.STRENGTH -> {
                when {
                    stallCount >= 6 -> "Consider changing your rep scheme or adding a variation of $exerciseName for 3-4 weeks, then re-test."
                    stallCount >= 4 -> "Try a deload week at 60% of your $exerciseName max, then build back with smaller jumps."
                    else -> "Add volume: try 3-5 extra sets of $exerciseName per week at 70-80% to break through."
                }
            }
            PlateauType.VOLUME -> {
                "Your $exerciseName volume has plateaued. Try adding 1-2 sets per session or increase weight by 5%."
            }
        }
    }

    /** Parse an ISO8601 date string to LocalDate. Tolerant of multiple ISO formats. */
    private fun parseDate(dateString: String): LocalDate {
        return try {
            LocalDate.parse(dateString.substring(0, 10))
        } catch (_: Exception) {
            try {
                LocalDateTime.parse(dateString).toLocalDate()
            } catch (_: Exception) {
                LocalDate.now()
            }
        }
    }
}
